sound/soc/sof/intel/hda-pcm.c

```diff
@@ -263,6 +263,25 @@ int hda_dsp_pcm_open(struct snd_sof_dev *sdev,
 		snd_pcm_hw_constraint_mask64(substream->runtime, SNDRV_PCM_HW_PARAM_FORMAT,
 					     SNDRV_PCM_FMTBIT_S16 | SNDRV_PCM_FMTBIT_S32);
 
+	/*
+	 * The dsp_buffer_time_ms is the length of the host DMA buffer of the
+	 * firmware.
+	 * On playback start the DMA will transfer dsp_buffer_time_ms amount of
+	 * data in one initial burst to fill up the host DMA buffer.
+	 * Consequent DMA burst sizes are shorter and their length can vary.
+	 * To make sure that userspace allocate large enough ALSA buffer we need
+	 * to place a constraint on the buffer time.
+	 *
+	 * Exact dsp_buffer_time_ms constraint is racy for user space in case of
+	 * large buffer, so set the constraint to a minimum of
+	 * 2x dsp_buffer_time_ms.
+	 */
+	if (spcm->stream[substream->stream].dsp_buffer_time_ms)
+		snd_pcm_hw_constraint_minmax(substream->runtime,
+			SNDRV_PCM_HW_PARAM_BUFFER_TIME,
+			spcm->stream[substream->stream].dsp_buffer_time_ms * USEC_PER_MSEC * 2,
+			UINT_MAX);
```

Comment:
By default 

````c
sps->dsp_buffer_time_ms = SOF_IPC4_MIN_DMA_BUFFER_SIZE;
````
so this forces the ALSA buffer to be a minimum of 4ms.

Also this is a bit ironic that the race happens for large buffers, IIRC the issue was an xrun with small periods with mplayer? If this is only for large buffers, could we enforce this constraint when 

````spcm->stream[substream->stream].dsp_buffer_time_ms > SOF_IPC4_MIN_DMA_BUFFER_SIZE;````

Bear with me, reviewing on Friday afternoon...

Reply:
The xrun happens with large buffers because application rarely if ever took small (< 4ms) buffers.

Reply:
so in practice could we not make this constraint conditional, e.g.

````c
if ( spcm->stream[substream->stream].dsp_buffer_time_ms > SOF_IPC4_MIN_DMA_BUFFER_SIZE) {
    snd_pcm_hw_constraint_minmax(substream->runtime,
			SNDRV_PCM_HW_PARAM_BUFFER_TIME,
			spcm->stream[substream->stream].dsp_buffer_time_ms * USEC_PER_MSEC * 2,
			UINT_MAX);
}

Reply:
> so in practice could we not make this constraint conditional, e.g.
> 
> ```c
> if ( spcm->stream[substream->stream].dsp_buffer_time_ms > SOF_IPC4_MIN_DMA_BUFFER_SIZE) {
>     snd_pcm_hw_constraint_minmax(substream->runtime,
> 			SNDRV_PCM_HW_PARAM_BUFFER_TIME,
> 			spcm->stream[substream->stream].dsp_buffer_time_ms * USEC_PER_MSEC * 2,
> 			UINT_MAX);
> }
> ```

No, the ALSA buffer must be constrained at all times, if the PCM uses the default 2ms buffer then the first DMA burst will copy 2ms data, the buffer should be double of that.

While on capture the ALSA buffer size should be at least 2ms, the DMA copies 1ms chunks from DSP to host, this I missed.